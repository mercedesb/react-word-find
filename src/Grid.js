export class Grid {
  constructor(grid=null) {
    this.grid = grid || new Array(this.rows * this.columns)
  }

  get size() {
    return this.grid.length
  }

  get rows() {
    return 7
  }

  get columns() {
    return 7
  }
  
  index(rowIndex, columnIndex) {
    if(!!columnIndex) {
      return rowIndex * this.columns + columnIndex
    }
    else {
      return rowIndex
    }
  }

  row(position) {
    return Math.trunc(position / 7)
  }

  column(position) {
    return position % 7
  }

  at(position) {
    return this.grid[position]
  }

  setAt(position, value) {
    this.grid[position] = value
  }

  clone() {
    return new Grid(this.grid)
  }

  fill() {
    const letters = 'abcdefghijklmnopqrstuvwxyz'.split('')
    for (var i = 0; i < this.size; i++ ) {
      if (!this.grid[i]) {
        this.grid[i] = letters[Math.floor((Math.random()*letters.length))]
      }
    }
  }
}