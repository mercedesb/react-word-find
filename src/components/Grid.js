export class Grid {
  constructor(length=7, grid=null) {
    this.length = length
    this.grid = grid || new Array(this.length * this.length)
  }

  get size() {
    return this.grid.length
  }

  index(rowIndex, columnIndex) {
    if(!!columnIndex) {
      return rowIndex * this.length + columnIndex
    }
    else {
      return rowIndex
    }
  }

  row(position) {
    return Math.trunc(position / 7)
  }

  column(position) {
    return position % 7
  }

  at(position) {
    return this.grid[position]
  }

  setAt(position, value) {
    this.grid[position] = value
  }

  clone() {
    return new Grid(this.length, this.grid)
  }

  fill() {
    const letters = 'abcdefghijklmnopqrstuvwxyz'.split('')
    for (var i = 0; i < this.size; i++ ) {
      if (!this.grid[i]) {
        this.grid[i] = letters[Math.floor((Math.random()*letters.length))]
      }
    }
  }

  print() {
    return this.grid
  }
}