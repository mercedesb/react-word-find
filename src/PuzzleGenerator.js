import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Grid } from './Grid'

const DIRECTIONS = [
    { // right
      moveRow: 0,
      moveColumn: 1
    },
    { // down
      moveRow: 1,
      moveColumn: 0
    }
  ]

export class PuzzleGenerator extends Component {
  state = {
    grid: {},
    noSolution: false
  }

  static generate = (wordBank) => {
    const sortedWordBank = wordBank.sort((a, b) => b.length - a.length)
    let grid = new Grid()
    const positions = Array.from({length: grid.size}, (value, index) => index)

    const configStack = [ PuzzleGenerator.getNewConfig(grid, sortedWordBank, positions) ]

    let currentConfig
    let direction
    let position

    while(true) {
      currentConfig = configStack[configStack.length - 1]
      if (!currentConfig) {
        return { grid: {}, noSolution: true }
      }

      direction = PuzzleGenerator.getDirection(currentConfig)
      position = currentConfig.positions[currentConfig.positions.length - 1]

      if (!position) {
        sortedWordBank.unshift(currentConfig.word)
        configStack.pop()
      }
      else {
        grid = PuzzleGenerator.tryWord(currentConfig.grid, currentConfig.word, position, direction)

        if (!!grid) {
          if (!!sortedWordBank.length) {
            configStack.push(PuzzleGenerator.getNewConfig(grid, sortedWordBank, positions))
          }
          else {
            // success!
            break
          }
        } 
      }
    }
    
    grid.fill()
    return { grid: grid, noSolution: false }
  }

  static getNewConfig = (grid, sortedWordBank, positions) => {
    return {
      grid: grid,
      word: sortedWordBank.shift(),
      directions: PuzzleGenerator.shuffle(DIRECTIONS),
      positions: PuzzleGenerator.shuffle(positions)
    }
  }

  static getDirection = (config) => {
    let direction = config.directions.pop()
    if (!direction) {
      config.positions.pop()
      config.directions = PuzzleGenerator.shuffle(DIRECTIONS)
      direction = config.directions.pop()
    }
    return direction
  }

  static tryWord = (grid, word, position, direction) => {
    const clone = grid.clone()
    const letters = word.split('')

    let currentRow = clone.row(position)
    let currentColumn = clone.column(position)
    let letter
    let currentIndex

    while (PuzzleGenerator.withinGrid(clone, currentRow, currentColumn)) {
      letter = letters.shift()
      if (!letter) {
        break
      }

      currentIndex = clone.index(currentRow, currentColumn)

      if (PuzzleGenerator.positionAvailable(clone, currentIndex, letter)) {
        clone.setAt(currentIndex, letter)
        currentRow = PuzzleGenerator.getNextRowPosition(direction, currentRow)
        currentColumn = PuzzleGenerator.getNextColumnPosition(direction, currentColumn)
      }
      else {
        return null
      }
    }
    return letters.length > 0 ? null : clone
  }

  static withinGrid = (grid, row, column) => {
    return row >= 0 && row < grid.rows && column >= 0 && column < grid.columns
  }

  static positionAvailable = (grid, position, letter) => {
    return !grid.at(position) || grid.at(position) == letter
  }

  static getNextRowPosition = (direction, row) => {
    return row += direction.moveRow
  }

  static getNextColumnPosition = (direction, column) => {
    return column += direction.moveColumn
  }

  static shuffle = (array) => {
    const clone = array.slice(0)
    for (let i = clone.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [clone[i], clone[j]] = [clone[j], clone[i]]; // eslint-disable-line no-param-reassign
    }
    return clone
  }

  static getDerivedStateFromProps = (props, state) => {
    const wordBank = props.wordBank.slice(0)
    return PuzzleGenerator.generate(wordBank)
  }

  getLetters = () => {
    return this.state.grid.print().map((letter, i) => {
      return <div className='WordPuzzle-letter' key={i}>{letter}</div>
    })
  }

  render () {
    return (
      <div className='WordPuzzle WordPuzzle--seven'>
        { this.state.noSolution && (
            <span>No solution, try again</span>
          )
        }
        { !this.state.noSolution && (
            this.getLetters()
          )
        }
      </div>
    )
  }
}

PuzzleGenerator.propTypes = {
  wordBank: PropTypes.array
}

