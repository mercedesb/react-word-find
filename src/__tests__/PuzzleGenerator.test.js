/* eslint-env jest */
import * as React from 'react'
import { shallow } from 'enzyme'
import { PuzzleGenerator } from '../PuzzleGenerator'
import { Grid } from '../Grid'

let subject 
let wordBank

function loadSubject (props = {}) {
  wordBank = ['hello', 'world']
  subject = shallow(<PuzzleGenerator
    wordBank={wordBank}
    {...props}
  />)
}

describe('#render', () => {
  it('renders properly', () => {
    loadSubject()
    expect(subject).toMatchSnapshot()
  })
})

describe('#generate', () => {
  describe('when no solution possible', () => {
    it('raises an error', () => {
      const wordBank = [
        'prudence',
        'dog',
        'jenny',
        'bird',
      ]
      loadSubject({wordBank})

      expect(() => {
        subject.instance().generate()
      }).toThrow()
    })
  })

  describe('when solution is possible', () => {
    it('sets the state', () => {
      const wordBank = [
        'pru',
        'magoo',
        'dog',
        'jenny',
        'bird'
      ]
      loadSubject({wordBank})
      subject.instance().generate()
      expect(subject.state('grid').size).toEqual(49)
    })
  })
})

describe('#getNewConfig', () => {
  it('returns an object with the expected shape', () => {
    loadSubject()
    const grid = new Grid()
    const wordBank = ['hello', 'there', 'hi']
    const positions = [0, 1, 2]
    const expectedObj = {
      grid: grid,
      word: 'hello',
      directions: expect.anything(),
      positions: expect.arrayContaining(positions)
    }
    expect(subject.instance().getNewConfig(grid, wordBank, positions)).toMatchObject(expectedObj)
  })
})

describe('#getDirection', () => {
  describe('with a direction in the stack', () => {
    it('returns a direction', () => {
      loadSubject()
      const grid = new Grid()
      const positions = [0, 1, 2]
      const right = {
        moveRow: 0,
        moveColumn: 1
      }

      const config = {
        grid,
        positions,
        directions: [right],
        word: 'hello'
      }

      expect(subject.instance().getDirection(config)).toEqual(right)
    })
  })

  describe('without a direction in the stack', () => {
    it('returns a direction', () => {
      loadSubject()
      const grid = new Grid()
      const positions = [0, 1, 2]
      const directions = []
      const config = {
        grid,
        directions,
        positions,
        word: 'hello'
      }

      expect(subject.instance().getDirection(config)).toBeDefined()
      expect(subject.instance().getDirection(config)).not.toBeNull()
    })
  })

  // TODO: what happens if no direction and no position 
})

describe('#tryWord', () => {
  describe('with a word that will fit within the row', () => {
    it('returns a new grid with the word placed', () => {
      loadSubject()
      const grid = new Grid()
      const word = 'dog'
      const position = 2
      const right = {
        moveRow: 0,
        moveColumn: 1
      }

      const returnedGrid = subject.instance().tryWord(grid, word, position, right)
      expect(returnedGrid).not.toBeNull()
      expect(returnedGrid.at(2)).toEqual('d')
      expect(returnedGrid.at(3)).toEqual('o')
      expect(returnedGrid.at(4)).toEqual('g')
    })
  })
  
  describe('with a word that will not fit within the row', () => {
    it('returns null', () => {
      loadSubject()
      const grid = new Grid()
      const word = 'prudence'
      const position = 2
      const right = {
        moveRow: 0,
        moveColumn: 1
      }
      expect(subject.instance().tryWord(grid, word, position, right)).toBeNull()
    })
  })

  describe('with a word that will fit within the column', () => {
    it('returns a new grid with the word placed', () => {
      loadSubject()
      const grid = new Grid()
      const word = 'dog'
      const position = 2
      const down = {
        moveRow: 1,
        moveColumn: 0
      }
      const returnedGrid = subject.instance().tryWord(grid, word, position, down)
      expect(returnedGrid).not.toBeNull()
      expect(returnedGrid.at(2)).toEqual('d')
      expect(returnedGrid.at(9)).toEqual('o')
      expect(returnedGrid.at(16)).toEqual('g')
    })
  })

  describe('with a word that will fit not within the column', () => {
    it('returns null', () => {
      loadSubject()
      const grid = new Grid()
      const word = 'prudence'
      const position = 2
      const down = {
        moveRow: 1,
        moveColumn: 0
      }
      expect(subject.instance().tryWord(grid, word, position, down)).toBeNull()
    })
  })
})

describe('#withinGrid', () => {
  describe('when rows and columns are inside grid', () => {
    it('returns true', () => {
      const grid = new Grid()
      expect(subject.instance().withinGrid(grid, 3, 3)).toEqual(true)
    })
  })

  describe('when rows is at edge of grid', () => {
    it('returns true', () => {
      const grid = new Grid()
      expect(subject.instance().withinGrid(grid, 0, 3)).toEqual(true)
    })
  })

  describe('when columns is at edge of grid', () => {
    it('returns true', () => {
      const grid = new Grid()
      expect(subject.instance().withinGrid(grid, 3, 6)).toEqual(true)
    })
  })

  describe('when rows are beyond grid', () => {
    it('returns false', () => {
      const grid = new Grid()
      expect(subject.instance().withinGrid(grid, 7, 3)).toEqual(false)
    })
  })

  describe('when rows are beyond grid', () => {
    it('returns false', () => {
      const grid = new Grid()
      expect(subject.instance().withinGrid(grid, 3, 9)).toEqual(false)
    })
  })
})

describe('#positionAvailable', () => {
  describe('when there is nothing at the current position', () => {
    it('returns true', () => {
      loadSubject()
      const grid = new Grid(['h', null, null, 'i' ])
      expect(subject.instance().positionAvailable(grid, 1, 'a')).toEqual(true)
    })
  })

  describe('when the same letter is at the current position', () => {
    it('returns true', () => {
      loadSubject()
      const grid = new Grid(['h', null, null, 'i' ])
      expect(subject.instance().positionAvailable(grid, 0, 'h')).toEqual(true)
    })
  })

  describe('when a different letter is at the current position', () => {
    it('returns true', () => {
      loadSubject()
      const grid = new Grid(['h', null, null, 'i' ])
      expect(subject.instance().positionAvailable(grid, 0, 'a')).toEqual(false)
    })
  })
})

describe('#getNextRowPosition', () => {
  describe('when direction is right', () => {
    it('returns same row index', () => {
      loadSubject()
      
      const right = {
        moveRow: 0,
        moveColumn: 1
      }

      expect(subject.instance().getNextRowPosition(right, 0)).toEqual(0)
    })
  })

  describe('when direction is down', () => {
    it('returns increased row index', () => {
      loadSubject()
        
      const down = { 
        moveRow: 1,
        moveColumn: 0
      }

      expect(subject.instance().getNextRowPosition(down, 0)).toEqual(1)
    })
  })
})

describe('#getNextColumnPosition', () => {
  describe('when direction is right', () => {
    it('returns increased column index', () => {
      loadSubject()
      
      const right = {
        moveRow: 0,
        moveColumn: 1
      }

      expect(subject.instance().getNextColumnPosition(right, 0)).toEqual(1)
    })
  })

  describe('when direction is down', () => {
    it('returns same column index', () => {
      loadSubject()
        
      const down = { 
        moveRow: 1,
        moveColumn: 0
      }

      expect(subject.instance().getNextColumnPosition(down, 0)).toEqual(0)
    })
  })
})

describe('#shuffle', () => {
  it('returns a new array', () => {
    loadSubject()
    expect(subject.instance().shuffle(wordBank)).not.toBe(wordBank)
  })

  it('returns an array with the same items', () => {
    loadSubject()
    expect(subject.instance().shuffle(wordBank)).toEqual(expect.arrayContaining(wordBank))
  })
})
