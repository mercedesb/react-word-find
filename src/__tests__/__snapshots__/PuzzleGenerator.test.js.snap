// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PuzzleGenerator
    wordBank={
        Array [
            "hello",
            "world",
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        "m",
        "v",
        "k",
        "s",
        "h",
        "w",
        "t",
        "z",
        "w",
        "x",
        "h",
        "e",
        "d",
        "c",
        "t",
        "o",
        "x",
        "a",
        "l",
        "j",
        "l",
        "o",
        "r",
        "i",
        "b",
        "l",
        "w",
        "o",
        "g",
        "l",
        "w",
        "o",
        "o",
        "s",
        "t",
        "u",
        "d",
        "o",
        "v",
        "j",
        "o",
        "w",
        "x",
        "f",
        "r",
        "v",
        "o",
        "w",
        "r",
      ],
    },
    "ref": null,
    "rendered": Array [
      "m",
      "v",
      "k",
      "s",
      "h",
      "w",
      "t",
      "z",
      "w",
      "x",
      "h",
      "e",
      "d",
      "c",
      "t",
      "o",
      "x",
      "a",
      "l",
      "j",
      "l",
      "o",
      "r",
      "i",
      "b",
      "l",
      "w",
      "o",
      "g",
      "l",
      "w",
      "o",
      "o",
      "s",
      "t",
      "u",
      "d",
      "o",
      "v",
      "j",
      "o",
      "w",
      "x",
      "f",
      "r",
      "v",
      "o",
      "w",
      "r",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "m",
          "v",
          "k",
          "s",
          "h",
          "w",
          "t",
          "z",
          "w",
          "x",
          "h",
          "e",
          "d",
          "c",
          "t",
          "o",
          "x",
          "a",
          "l",
          "j",
          "l",
          "o",
          "r",
          "i",
          "b",
          "l",
          "w",
          "o",
          "g",
          "l",
          "w",
          "o",
          "o",
          "s",
          "t",
          "u",
          "d",
          "o",
          "v",
          "j",
          "o",
          "w",
          "x",
          "f",
          "r",
          "v",
          "o",
          "w",
          "r",
        ],
      },
      "ref": null,
      "rendered": Array [
        "m",
        "v",
        "k",
        "s",
        "h",
        "w",
        "t",
        "z",
        "w",
        "x",
        "h",
        "e",
        "d",
        "c",
        "t",
        "o",
        "x",
        "a",
        "l",
        "j",
        "l",
        "o",
        "r",
        "i",
        "b",
        "l",
        "w",
        "o",
        "g",
        "l",
        "w",
        "o",
        "o",
        "s",
        "t",
        "u",
        "d",
        "o",
        "v",
        "j",
        "o",
        "w",
        "x",
        "f",
        "r",
        "v",
        "o",
        "w",
        "r",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
